core/src/main/scala/caliban/wrappers/Caching.scala

```diff
@@ -0,0 +1,329 @@
+package caliban.wrappers
+
+import caliban.ResponseValue.ObjectValue
+import caliban.Value._
+import caliban._
+import caliban.execution.{ ExecutionRequest, Field, FieldInfo }
+import caliban.introspection.adt._
+import caliban.parsing.adt.{ Directive, Document }
+import caliban.schema.Annotations.{ GQLDirective, GQLName }
+import caliban.schema.Types
+import caliban.wrappers.Wrapper.{ EffectfulWrapper, FieldWrapper, OverallWrapper, ValidationWrapper }
+import zio.prelude._
+import zio.query.ZQuery
+import zio.{ durationInt, Duration, FiberRef, Ref, UIO, Unsafe, ZIO }
+
+import java.util.concurrent.{ ConcurrentHashMap, TimeUnit }
+
+object Caching {
+  val DirectiveName     = "cacheControl"
+  val MaxAgeName        = "maxAge"
+  val ScopeName         = "scope"
+  val InheritMaxAgeName = "inheritMaxAge"
+
+  private val cacheOverride: FiberRef[Option[CacheHint]] = Unsafe.unsafe(implicit u => FiberRef.unsafe.make(None))
+
+  private val cacheDirectives = List(
+    __Directive(
+      name = DirectiveName,
+      description = None,
+      locations = Set(__DirectiveLocation.FIELD_DEFINITION, __DirectiveLocation.OBJECT),
+      args = List(
+        __InputValue(
+          name = MaxAgeName,
+          None,
+          `type` = () => Types.int,
+          defaultValue = None
+        ),
+        __InputValue(
+          name = ScopeName,
+          None,
+          `type` = () => CacheScope._type,
+          defaultValue = None
+        ),
+        __InputValue(
+          name = InheritMaxAgeName,
+          None,
+          `type` = () => Types.boolean,
+          defaultValue = None
+        )
+      ),
+      isRepeatable = false
+    )
+  )
+
+  def aspect[R](output: Ref[CachePolicy] => Wrapper[R]): GraphQLAspect[Nothing, R] =
+    new Default(output)
```

Comment:
Abstracted this out, this is to support a fast follow where I'd like to support full response caching within caliban.

Reply:
Oh that sounds interesting!